### ISSUE TITLE
Release trace context level-1

### ISSUE THREAD

**Issue by dyladan** [w3c/distributed-tracing#469]
Since the most recent level-1 publish, there has been an erratum raised and fixed.

It might be good to wait for #465 to merge first so this only needs to be done once.

**Comment by plehegar** [comment]
[Diff between the Rec and the latest level-1](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fwww.w3.org%2FTR%2Ftrace-context%2F&doc2=https%3A%2F%2Flabs.w3.org%2Fspec-generator%2F%3Ftype%3Drespec%26url%3Dhttps%253A%252F%252Fraw.githack.com%252Fw3c%252Ftrace-context%252Flevel-1%252Findex.html%253FspecStatus%253DWD%26file%3D)

**Comment by plehegar** [comment]
A more [useful diff](https://github.com/w3c/trace-context/compare/c190b70e3b18fc5a61a090aa319f5e9c2f211316...level-1)

**Comment by plehegar** [comment]
1. All changes to the BNF are just about fixing the notation https://github.com/w3c/trace-context/pull/432
2. https://github.com/w3c/trace-context/pull/465 removes some confusing vendor/tenant wording
3. https://github.com/w3c/trace-context/issues/428 fixes a contradiction between the mutation. In effect, it changes a MUST to a SHOULD on how keys are ordered if there is a mutation.

All of the changes do not affect implementations

**Comment by swickr** [comment]
Re: #428 -- How likely is it that some implementation relied on moving modified keys to the left?  How many implementations consistently moved a modified key?

**Comment by basti1302** [comment]
@swickr
> How likely is it that some implementation relied on moving modified keys to the left?

Let me rephrase this question slightly: How likely is it that an implementation B relied on an implementation A (where A is upstream of B in that particular transaction) to move the key that A modifies to the left. That would be the only situation in which the change from MUST to SHOULD could be considered a breaking change and I assume that is the intent behind your question, correct?

IMO that would be an extremely unlikely scenario, I would even say impossible. The `tracestate` header field value is a set of _vendor specific_ key value pairs. Each vendor is supposed to only care about the key value pairs which it adds to `tracestate` _itself_. Everything else in there is to be treated as a list of opaque key value pairs. Thus, an implementation is not supposed to derive any information from key value pairs from other vendors. In my understanding, this also implies that an implementation is not supposed to analyze the relative positions of key value pairs that belong to other implementations.

> How many implementations consistently moved a modified key?

I do not understand the motivation for this question. If a hypothetical implementation existed that does not consistently move they key value pair to the left when updating it, it would be in violation of the spec _now_ and technically, it would no longer violate the spec _after_ this update.  

FWIW, to the best of my knowledge, there is no such implementation. All implementations I know of that modify tracestate work like this:
1. Check if their own key is present in tracestate.
2. If it is already present, update it and move it to the left-most position when propagating `tracestate` downsteam.
3. If it is not present, simply prepend their own key value pair as the left-most position when propagating `tracestate` downsteam.

**Comment by plehegar** [comment]
In [section 3.5 Mutating the tracestate Field (level-1 branch)](https://raw.githack.com/w3c/trace-context/level-1/index.html?specStatus=WD#mutating-the-tracestate-field), I read:

* "Modified keys MUST be moved to the beginning (left) of the list." (first paragraph)
* Update an existing value. [...]. "Modified keys SHOULD be moved to the beginning (left) of the list."

So, I would conclude that the change from MUST to SHOULD in the second paragraph did not affect the fact that implementations are still expected to move the modified key to the beginning (left) of the list. Shouldn't we modify the first sentence as well?

**Comment by dyladan** [comment]
> IMO that would be an extremely unlikely scenario, I would even say impossible. The `tracestate` header field value is a set of _vendor specific_ key value pairs. Each vendor is supposed to only care about the key value pairs which it adds to `tracestate` _itself_. Everything else in there is to be treated as a list of opaque key value pairs. Thus, an implementation is not supposed to derive any information from key value pairs from other vendors. In my understanding, this also implies that an implementation is not supposed to analyze the relative positions of key value pairs that belong to other implementations.

I agree except on one point. It is important that entries are moved left if the header ever grows too long and needs to be truncated. This ensures the most recently updated keys are the ones which are kept after truncation.

**Comment by basti1302** [comment]
> I agree except on one point. It is important that entries are moved left if the header ever grows too long and needs to be truncated. This ensures the most recently updated keys are the ones which are kept after truncation.

Yes, sure, but that does not apply directly to "How likely is it that some implementation specifically relied on moving modified keys to the left?", does it?

The length/truncation argument could be used as an argument to keep the MUST and not go with the SHOULD in the first place. But that is not what we decided, I think we don't want to rehash the MUST vs. SHOULD discussion here, do we? But if we feel that the SHOULD is strong enough (for level 2) then I don't see how the length/truncation argument would relate to the question whether a level 1 implementation would specifically rely on other implementations moving key-value pairs around or not.

**Comment by plehegar** [comment]
Publication of the updated Rec has been requested for November 23.

**Comment by plehegar** [comment]
https://www.w3.org/TR/2021/REC-trace-context-1-20211123/

This issue can be closed.
